### DISCUSSION TITLE
타입가드란?

### DISCUSSION THREAD

**Discussion by skyhanull** [wanted-9th-3team/frontend-study#44] [프론트 기술 질문 공간]
타입스크립트의 타입가드에 대해서 설명해주세요

**Comment by michoball** [comment]
런타임 환경에서 변수의 타입을 검사해서 특정한 타입일 경우에만 해당 타입에 대한 작업을 수행할 수 있도록 해주는 코드입니다. 

typeof 나 instanceof 같은 것을 사용해서 가드를 할 수도 있고 따로 함수로 구현할 수도 있습니다. 

타입가드를 사용하면 코드의 가독성과 유지 보수성을 높일 수 있고 런타임오류를 방지할 수 있다는 장점이 많아 잘 사용할 수 있도록 하는 것이 좋습니다.

**Comment by WooGie911** [comment]
타입 가드(Type Guard)는 컴파일러가 타입을 예측할 수 있도록 코드를 작성해서 버그가 발생하지 않도록 예방하는 방법입니다. 

정확하게 표현하면 TypeScript 컴파일러가 타입 추론(Type Inference)을 할 수 있도록 하는 방법입니다.

흔히 typeof, instanceof 뿐만 아니라 등호연산자와 in 연산자를 사용합니다.

**Comment by skyhanull** [comment]
타입스크립트의 타입가드란 데이터의 타입을 알 수 없거나, 될 수 있는 타입이 여러 개라고 가정할 때 조건문을 통해 데이터의 타입을 좁혀나가는 것입니다.
typeof, instanceof를 사용하여 데이터의 타입에 따라 대응하여 에러를 최소화할 수 있고, 타입을 통해 '가드'하는 코드, 즉 방어적인 코드를 짤 수 있습니다.

**Comment by javascriptbiggosoo** [comment]
자바스크립트에서 타입안정성을 높이기 위해 `isArray`, `instanceOf` 등 많은 장치들이 있지만 이것들은 옵셔널으로 강제성이 없습니다. 타입스크립트는 디폴트로 타입가드를 설정해주어야하며 원하지 않을경우 `any`를 지정해줘야하는 옵트아웃 방식입니다.

**Comment by BaekYuri** [comment]
타입가드를 사용하면 조건문에서 객체의 타입을 좁혀나갈 수 있습니다. typeof, instanceof, in, 리터럴 가드 등을 사용하여 타입을 좁혀나갈 수 있고, 여러가지 타입이 올 수 있는 상황에서 타입을 좁혀나가면서 안정성을 올릴 수 있습니다.
